base/linalg/rowvector.jl

```diff
@@ -0,0 +1,183 @@
+
+# similar()
+@inline similar(rowvec::RowVector) = RowVector(similar(rowvec.vec))
+@inline similar{T}(rowvec::RowVector, ::Type{T}) = RowVector(similar(rowvec.vec, transpose_type(T)))
+# There is no resizing similar() because it would be ambiguous if the result were a Matrix or a RowVector
+
+# Basic methods
+@inline transpose(vec::AbstractVector) = RowVector(vec)
+@inline ctranspose{T}(vec::AbstractVector{T}) = RowVector(conj(vec))
+@inline ctranspose{T<:Real}(vec::AbstractVector{T}) = RowVector(vec)
+
+@inline transpose(rowvec::RowVector) = rowvec.vec
+@inline ctranspose{T}(rowvec::RowVector{T}) = conj(rowvec.vec)
+@inline ctranspose{T<:Real}(rowvec::RowVector{T}) = rowvec.vec
+
+# Some overloads from Base
+@inline transpose(r::Range) = RowVector(r)
+@inline ctranspose(r::Range) = RowVector(conj(r)) # is there such a thing as a complex range?
+
+# Strictly, these are unnecessary but will make things stabler if we introduce
+# a "view" for conj(::AbstractArray)
+@inline conj(rowvec::RowVector) = RowVector(conj(rowvec.vec))
+@inline conj{T<:Real}(rowvec::RowVector{T}) = rowvec
+
+# AbstractArray interface
+@inline length(rowvec::RowVector) =  length(rowvec.vec)
+@inline size(rowvec::RowVector) = (1, length(rowvec.vec))
+@inline size(rowvec::RowVector, d) = ifelse(d==2, length(rowvec.vec), 1)
+linearindexing{V<:RowVector}(::Union{V,Type{V}}) = LinearFast()
+
+@propagate_inbounds getindex(rowvec::RowVector, i) = transpose(rowvec.vec[i])
+@propagate_inbounds setindex!(rowvec::RowVector, v, i) = setindex!(rowvec.vec, transpose(v), i)
+
+# Cartesian indexing is distorted by getindex
+# Furthermore, Cartesian indexes don't have to match shape, apparently!
+@inline function getindex(rowvec::RowVector, i::CartesianIndex)
+    @boundscheck if !(i.I[1] == 1 && i.I[2] ∈ indices(rowvec.vec)[1] && check_tail_indices(i.I...))
+        throw(BoundsError(rowvec, i.I))
+    end
+    @inbounds return transpose(rowvec.vec[i.I[2]])
+end
+@inline function setindex!(rowvec::RowVector, v, i::CartesianIndex)
+    @boundscheck if !(i.I[1] == 1 && i.I[2] ∈ indices(rowvec.vec)[1] && check_tail_indices(i.I...))
+        throw(BoundsError(rowvec, i.I))
+    end
+    @inbounds rowvec.vec[i.I[2]] = transpose(v)
+end
+
+@propagate_inbounds getindex(rowvec::RowVector, ::CartesianIndex{0}) = getindex(rowvec)
+@propagate_inbounds getindex(rowvec::RowVector, i::CartesianIndex{1}) = getindex(rowvec, i.I[1])
+
+@propagate_inbounds setindex!(rowvec::RowVector, v, ::CartesianIndex{0}) = setindex!(rowvec, v)
+@propagate_inbounds setindex!(rowvec::RowVector, v, i::CartesianIndex{1}) = setindex!(rowvec, v, i.I[1])
+
+@inline check_tail_indices(i1, i2) = true
+@inline check_tail_indices(i1, i2, i3, is...) = i3 == 1 ? check_tail_indices(i1, i2, is...) :  false
+
+# Some conversions
+convert(::Type{AbstractVector}, rowvec::RowVector) = rowvec.vec
```

Comment:
Are we sure we want this? There aren't currently any conversions between arrays with different numbers of dimensions. If there were, I suspect we'd only allow converting `(n,)`-size arrays to `(n,1)`, not `(1,n)`-size arrays to `(n,)`.